dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,209 @@
+ *
+ *                                 [[1, 1, 7, 6],
+ *                                  [8, 1, 5, 7]],
+ *
+ *                                 [[6, 8, 8, 4],
+ *                                  [6, 2, 7, 5]]]
+ * shape = (3, 2, 4)
+ *
+ * @tparam new_axis if true, STACK mode is applied.
+ */
+template<typename T, bool new_axis = true, int dims = -1>
+struct TensorJoinCpu {
+  ///@{
+  /**
+   * @param in_shapes Shapes of input tensors.
+   * @param axis Axis, along which tensors will be joined. Accepts positive and negative values
+   *             (e.g. axis: -1 <=> ndims-1)
+   */
+  KernelRequirements Setup(KernelContext &ctx, span<const TensorShape<dims>> in_shapes, int axis) {
+    n_input_tensors_ = in_shapes.size();
+    auto ndims = in_shapes[0].sample_dim();
+    DALI_ENFORCE(axis >= -ndims + !new_axis && axis <= ndims - !new_axis,
+                 make_string("Incorrect axis. Actual: ", axis, ". Expected in [",
+                             -ndims + !new_axis, ", ", ndims - !new_axis, "] interval (",
+                             new_axis ? "STACK" : "CONCAT", " mode)"));
+    axis_ = axis >= 0 ? axis : ndims + axis + new_axis;
+
+    {
+      const auto &ref = in_shapes[0];
+      for (int i = 1; i < n_input_tensors_; i++) {
+        DALI_ENFORCE(in_shapes[i].sample_dim() == ref.sample_dim(),
+                     "Every input shape must have the same number of dimensions.");
+        for (int j = 0; j < ref.size(); j++) {
+          if (!new_axis) {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j] || j == axis_, make_string(
+                    "Number of samples in every dimension (except the concatenated one) "
+                    "must be the same (CONCAT mode). 0-th shape at index ", j, " has dimension ",
+                    ref.shape[j], ", while ", i, "-th shape at index ", j,
+                    " has dimension ", in_shapes[i][j]));
+          } else {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j], make_string(
+                    "Number of samples in every dimension must be the same (STACK mode). "
+                    "0-th shape at index ", j, " has dimension ", ref.shape[j], ", while ", i,
+                    "-th shape at index ", j, " has dimension ", in_shapes[i][j]));
+          }
+        }
+      }
+    }
+
+    KernelRequirements kr;
+    output_shape_ = DetermineShape<new_axis>(in_shapes, axis_);
+    kr.output_shapes.resize(1);
+    TensorListShape<> tmp({output_shape_});  // clang's destructor bug still haunting
+    kr.output_shapes[0] = tmp;
+    return kr;
+  }
+
+
+  KernelRequirements Setup(KernelContext &ctx, span<const InTensorCPU<T, dims>> in, int axis) {
+    std::vector<TensorShape<>> in_shapes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      in_shapes[i] = in[i].shape;
+    }
+    return Setup(ctx, make_span(in_shapes), axis);
+  }
+  ///@}
+
+  static constexpr int output_dims = (dims == DynamicDimensions ? DynamicDimensions :
```

Comment:
I'd move this definition to the top (line 109)

Reply:
I've intentionally put it here, sufficing "put you definition closest to the using point" policy

Reply:
Normally we don't write definitions in between member functions. IMHO it makes this definition hard to find. Anyway, not pushing.